solang-parser/src/pt.rs

```diff
@@ -1,1003 +1,1688 @@
+    /// The contract type.
     pub ty: ContractTy,
+    /// The identifier.
+    ///
+    /// This field is `None` only if an error occurred during parsing.
     pub name: Option<Identifier>,
+    /// The list of inheritance specifiers.
     pub base: Vec<Base>,
+    /// The list of contract parts.
     pub parts: Vec<ContractPart>,
 }
 
+/// An event parameter.
+///
+/// `<ty> [indexed] [name]`
 #[derive(Debug, PartialEq, Eq, Clone)]
 #[cfg_attr(feature = "pt-serde", derive(Serialize, Deserialize))]
 pub struct EventParameter {
-    pub ty: Expression,
+    /// The code location.
     pub loc: Loc,
+    /// The type.
+    pub ty: Expression,
+    /// Whether this parameter is indexed.
     pub indexed: bool,
+    /// The optional identifier.
     pub name: Option<Identifier>,
 }
 
+/// An event definition.
+///
+/// `event <name>(<fields>,*) [anonymous];`
 #[derive(Debug, PartialEq, Eq, Clone)]
 #[cfg_attr(feature = "pt-serde", derive(Serialize, Deserialize))]
 pub struct EventDefinition {
+    /// The code location.
     pub loc: Loc,
+    /// The identifier.
+    ///
+    /// This field is `None` only if an error occurred during parsing.
     pub name: Option<Identifier>,
+    /// The list of event parameters.
     pub fields: Vec<EventParameter>,
+    /// Whether this event is anonymous.
     pub anonymous: bool,
 }
 
+/// An error parameter.
+///
+/// `<ty> [name]`
 #[derive(Debug, PartialEq, Eq, Clone)]
 #[cfg_attr(feature = "pt-serde", derive(Serialize, Deserialize))]
 pub struct ErrorParameter {
-    pub ty: Expression,
+    /// The code location.
     pub loc: Loc,
+    /// The type.
+    pub ty: Expression,
+    /// The optional identifier.
     pub name: Option<Identifier>,
 }
 
+/// An error definition.
+///
+/// `error <name> (<fields>,*);`
 #[derive(Debug, PartialEq, Eq, Clone)]
 #[cfg_attr(feature = "pt-serde", derive(Serialize, Deserialize))]
 pub struct ErrorDefinition {
+    /// The code location.
     pub loc: Loc,
+    /// The `error` keyword.
     pub keyword: Expression,
+    /// The identifier.
+    ///
+    /// This field is `None` only if an error occurred during parsing.
     pub name: Option<Identifier>,
+    /// The list of error parameters.
     pub fields: Vec<ErrorParameter>,
 }
 
+/// An enum definition.
+///
+/// `enum <name> { <values>,* }`
 #[derive(Debug, PartialEq, Eq, Clone)]
 #[cfg_attr(feature = "pt-serde", derive(Serialize, Deserialize))]
 pub struct EnumDefinition {
+    /// The code location.
     pub loc: Loc,
+    /// The identifier.
+    ///
+    /// This field is `None` only if an error occurred during parsing.
     pub name: Option<Identifier>,
+    /// The list of values.
+    ///
+    /// This field contains `None` only if an error occurred during parsing.
     pub values: Vec<Option<Identifier>>,
 }
 
+/// A variable attribute.
 #[derive(Debug, PartialEq, Eq, Clone)]
 #[cfg_attr(feature = "pt-serde", derive(Serialize, Deserialize))]
+#[repr(u8)] // for cmp; order of variants is important
```

Comment:
Why is this necessary?

Reply:
The problem with using the derive macro is that it imposes the trait bound on all of the fields, which will also be checked during the call.
For example, deriving `PartialOrd` for `FunctionAttribute` transitively requires the trait to be implemented for `Mutability`, `IdentifierPath`, and `Base`; we can achieve this by implementing it manually, which goes against the whole point of deriving in the first place, or deriving it for these types too, which will eventually require deriving for almost the entire `pt` module.

By doing it manually we can make ordering depend only on the discriminant, regardless of the variants' value/fields; I think is fine for the few types (attributes, visibility) that have a recommended order in the Solidity standards.